Mathlib/Control/Monad/Freer.lean

```diff
@@ -0,0 +1,352 @@
+-/
+def run {ω : Type u} [Monoid ω] {α} : FreerWriter ω α → α × ω
+  | .pure a => (a, 1)
+  | .impure _ (WriterF.tell w) k =>
+      let (a, w') := run (k PUnit.unit)
+      (a, w * w')
+
+/--
+`listen` captures the log produced by a subcomputation incrementally. It traverses the computation,
+emitting log entries as encountered, and returns the accumulated log as a result.
+-/
+def listen {ω : Type u} [Monoid ω] {α : Type v} : FreerWriter ω α → FreerWriter ω (α × ω)
+  | .pure a => .pure (a, 1)
+  | .impure _ (WriterF.tell w) k =>
+      Freer.impure _ (WriterF.tell w) fun _ =>
+        listen (k PUnit.unit) >>= fun (a, w') =>
+          pure (a, w * w')
+
+/--
+`pass` allows a subcomputation to modify its own log. After traversing the computation and
+accumulating its log, the resulting function is applied to rewrite the accumulated log
+before re-emission.
+-/
+def pass {ω : Type u} [Monoid ω] {α} (m : FreerWriter ω (α × (ω → ω))) : FreerWriter ω α :=
+  let ((a, f), w) := run m
+  Freer.impure _ (WriterF.tell (f w)) (fun _ => .pure a)
+
+instance {ω : Type u} [Monoid ω] : MonadWriter ω (FreerWriter ω) where
+  tell w := Freer.impure _ (WriterF.tell w) (fun _ => .pure PUnit.unit)
+  listen := listen
+  pass := pass
+
+/--
+Evaluate a writer computation, returning the final result and discarding the log.
+-/
+def eval {ω : Type u} [Monoid ω] {α : Type v} (comp : FreerWriter ω α) : α :=
+  (run comp).1
+
+/--
+Execute a writer computation, returning only the accumulated log and discarding the result.
+-/
+def exec {ω : Type u} {α : Type v} [Monoid ω] (comp : FreerWriter ω α) : ω :=
+  (run comp).2
+
+end FreerWriter
+
+
+/-! ### Continuation Monad via `Freer` -/
+
+/-- Type constructor for continuation operations. -/
+inductive ContF (r : Type u) (α : Type v) where
+  /-- Call with current continuation: provides access to the current continuation. -/
+  | callCC : ((α → r) → r) → ContF r α
+
+instance {r : Type u} : Functor (ContF r) where
+  map f
+  | .callCC g => .callCC (fun k => g (k ∘ f))
+
+/-- Continuation monad via the `Freer` monad. -/
+abbrev FreerCont (r : Type u) := Freer (ContF r)
+
+namespace FreerCont
+
+instance {r : Type u} : Monad (FreerCont r) := inferInstance
+instance {r : Type u} : LawfulMonad (FreerCont r) := inferInstance
+
+/-- Run a continuation computation with the given continuation. -/
+def run {r : Type u} {α : Type v} : FreerCont r α → (α → r) → r
+  | .pure a, k => k a
+  | .impure _ (ContF.callCC g) cont, k => g (fun a => run (cont a) k)
+
+/-- Call with current continuation for the Freer continuation monad. -/
+def callCC {r : Type u} {α β : Type v} (f : MonadCont.Label α (FreerCont r) β → FreerCont r α) :
+FreerCont r α :=
+  Freer.impure _ (ContF.callCC (fun k =>
+    run (f ⟨fun x => Freer.impure _ (ContF.callCC (fun _ => k x)) Freer.pure⟩) k
+  )) Freer.pure
+
+instance {r : Type u} : MonadCont (FreerCont r) where
+  callCC := FreerCont.callCC
+
+end FreerCont
+
+/-! ### Basic Examples -/
```

Comment:
I think these belong in a test file.

Can you make them provable by `simp` instead of `rfl`, by adding suitable `simp` lemmas?